File: app_pojavlauncher/src/main/java/net/kdt/pojavlaunch/modloaders/BTADownloadTask.java
package net.kdt.pojavlaunch.modloaders;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Log;

import com.kdt.mcgui.ProgressLayout;

import git.artdeell.mojo.R;
import net.kdt.pojavlaunch.Tools;
import net.kdt.pojavlaunch.instances.Instance;
import net.kdt.pojavlaunch.instances.Instances;
import net.kdt.pojavlaunch.progresskeeper.ProgressKeeper;
import net.kdt.pojavlaunch.utils.FileUtils;

import java.io.File;
import java.io.IOException;
import java.net.URL;

public class BTADownloadTask implements Runnable {
    private static final String BASE_JSON = "{\"inheritsFrom\":\"b1.7.3\",\"mainClass\":\"net.minecraft.client.Minecraft\",\"libraries\":[{\"name\":\"bta-client:bta-client:%1$s\",\"downloads\":{\"artifact\":{\"path\":\"bta-client/bta-client-%1$s.jar\",\"url\":\"%2$s\"}}}],\"id\":\"%3$s\"}";
    private static final String BASE_JSON_LWJGL3 = "{\"inheritsFrom\":\"b1.7.3\",\"mainClass\":\"net.minecraft.client.Minecraft\",\"excludedLibraries\":[\"org.lwjgl.lwjgl:lwjgl\",\"org.lwjgl.lwjgl:lwjgl_util\",\"org.lwjgl.lwjgl:lwjgl-platform\"],\"libraries\":[{\"name\":\"bta-client:bta-client:%1$s\",\"downloads\":{\"artifact\":{\"path\":\"bta-client/bta-client-%1$s.jar\",\"url\":\"%2$s\"}}},{\"name\":\"org.lwjgl:lwjgl:%4$s\",\"natives\":{\"android-arm\":\"natives-linux-arm32\",\"android-arm64\":\"natives-linux-arm64\",\"android-x86\":\"natives-linux-x86\",\"android-x86_64\":\"natives-linux\"}},{\"name\":\"org.lwjgl:lwjgl-glfw:%4$s\"},{\"name\":\"org.lwjgl:lwjgl-openal:%4$s\"},{\"name\":\"org.lwjgl:lwjgl-opengl:%4$s\",\"natives\":{\"android-arm\":\"natives-linux-arm32\",\"android-arm64\":\"natives-linux-arm64\",\"android-x86\":\"natives-linux-x86\",\"android-x86_64\":\"natives-linux\"}},{\"name\":\"org.lwjgl:lwjgl-stb:%4$s\",\"natives\":{\"android-arm\":\"natives-linux-arm32\",\"android-arm64\":\"natives-linux-arm64\",\"android-x86\":\"natives-linux-x86\",\"android-x86_64\":\"natives-linux\"}},{\"name\":\"org.lwjgl:lwjgl-tinyfd:%4$s\",\"natives\":{\"android-arm\":\"natives-linux-arm32\",\"android-arm64\":\"natives-linux-arm64\",\"android-x86\":\"natives-linux-x86\",\"android-x86_64\":\"natives-linux\"}}],\"id\":\"%3$s\"}";
    private static final String LWJGL3_VERSION = "3.3.3";
    private final ModloaderDownloadListener mListener;
    private final BTAUtils.BTAVersion mBtaVersion;

    public BTADownloadTask(ModloaderDownloadListener mListener, BTAUtils.BTAVersion mBtaVersion) {
        this.mListener = mListener;
        this.mBtaVersion = mBtaVersion;
    }

    @Override
    public void run() {
        ProgressKeeper.submitProgress(ProgressLayout.INSTALL_MODPACK, 0, R.string.fabric_dl_progress, "BTA");
        try {
            runCatching() ;
            mListener.onDownloadFinished(null);
        }catch (IOException e) {
            mListener.onDownloadError(e);
        }
        ProgressLayout.clearProgress(ProgressLayout.INSTALL_MODPACK);
    }

    private void tryDownloadIcon(Instance targetInstance) {
        try {
            Bitmap iconBitmap = BitmapFactory.decodeStream(new URL(mBtaVersion.iconUrl).openStream());
            targetInstance.encodeNewIcon(iconBitmap);
        }catch (IOException e) {
            Log.w("BTADownloadTask", "Failed to download bta icon", e);
        }
    }

    private void createJson(String btaVersionId) throws IOException {
        int[] intVersion; try {
            intVersion = BTAUtils.parseBTAVersion(mBtaVersion);
        } catch (NumberFormatException e){
            intVersion = new int[]{7, 3, 0};
        }
        // BTA 7.3+ requires LWJGL3
        Log.i("BTADownloadTask", "Installing BTA " + intVersion[0] + "." + intVersion[1] + "." + intVersion[2] + ", nightly: " + BTAUtils.isNightlyVersion(mBtaVersion));
        String btaJson = (intVersion[0] >= 7 && intVersion[1] >= 3) || BTAUtils.isNightlyVersion(mBtaVersion) ?
                String.format(BASE_JSON_LWJGL3, mBtaVersion.versionName, mBtaVersion.downloadUrl, btaVersionId, LWJGL3_VERSION) :
                String.format(BASE_JSON, mBtaVersion.versionName, mBtaVersion.downloadUrl, btaVersionId);
        File jsonDir = new File(Tools.DIR_HOME_VERSION, btaVersionId);
        File jsonFile = new File(jsonDir, btaVersionId+".json");
        FileUtils.ensureDirectory(jsonDir);
        Tools.write(jsonFile, btaJson);
    }

    // BTA doesn't have SHA1 checksums in its repositories, so the user may try to reinstall it
    // if it didn't work due to a broken download. So, for reinstalls like that to work,
    // we need to delete the old client jar to force the download of a new one.
    private void removeOldClient() throws IOException{
        File btaClientPath = new File(Tools.DIR_HOME_LIBRARY, String.format("bta-client/bta-client-%1$s.jar", mBtaVersion.versionName));
        if(btaClientPath.exists() && !btaClientPath.delete())
            throw new IOException("Failed to delete old client jar");
    }

    private void createProfile(String btaVersionId) throws IOException {
        Instance instance = Instances.createInstance(i -> {
            i.versionId = btaVersionId;
            i.name = "Better than Adventure!";
        }, "BTA-"+btaVersionId);
        tryDownloadIcon(instance);
    }

    public void runCatching() throws IOException {
        removeOldClient();
        String btaVersionId = "bta-"+mBtaVersion.versionName;
        createJson(btaVersionId);
        createProfile(btaVersionId);
    }
}

File: app_pojavlauncher/src/main/java/net/kdt/pojavlaunch/modloaders/BTADownloadTask.kt
package net.kdt.pojavlaunch.modloaders

import android.graphics.BitmapFactory
import android.util.Log
import com.kdt.mcgui.ProgressLayout
import git.artdeell.mojo.R
import net.kdt.pojavlaunch.Tools
import net.kdt.pojavlaunch.instances.Instance
import net.kdt.pojavlaunch.instances.Instances
import net.kdt.pojavlaunch.progresskeeper.ProgressKeeper
import net.kdt.pojavlaunch.utils.FileUtils
import java.io.File
import java.io.IOException
import java.net.URL

class BTADownloadTask(
    private val mListener: ModloaderDownloadListener,
    private val mBtaVersion: BTAUtils.BTAVersion
) : Runnable {
    override fun run() {
        ProgressKeeper.submitProgress(ProgressLayout.INSTALL_MODPACK, 0, R.string.fabric_dl_progress, "BTA")
        try {
            runCatching()
            mListener.onDownloadFinished(null)
        } catch (e: IOException) {
            mListener.onDownloadError(e)
        }
        ProgressLayout.clearProgress(ProgressLayout.INSTALL_MODPACK)
    }

    private fun tryDownloadIcon(targetInstance: Instance) {
        try {
            val iconBitmap = BitmapFactory.decodeStream(URL(mBtaVersion.iconUrl).openStream())
            targetInstance.encodeNewIcon(iconBitmap)
        } catch (e: IOException) {
            Log.w("BTADownloadTask", "Failed to download bta icon", e)
        }
    }

    @Throws(IOException::class)
    private fun createJson(btaVersionId: String) {
        val btaJson = String.format(BASE_JSON, mBtaVersion.versionName, mBtaVersion.downloadUrl, btaVersionId)
        val jsonDir = File(Tools.DIR_HOME_VERSION, btaVersionId)
        val jsonFile = File(jsonDir, "$btaVersionId.json")
        FileUtils.ensureDirectory(jsonDir)
        Tools.write(jsonFile, btaJson)
    }

    @Throws(IOException::class)
    private fun removeOldClient() {
        val btaClientPath = File(Tools.DIR_HOME_LIBRARY, String.format("bta-client/bta-client-%1\$s.jar", mBtaVersion.versionName))
        if (btaClientPath.exists() && !btaClientPath.delete()) throw IOException("Failed to delete old client jar")
    }

    @Throws(IOException::class)
    private fun createProfile(btaVersionId: String) {
        val instance = Instances.createInstance({ i ->
            i.versionId = btaVersionId
            i.name = "Better than Adventure!"
        }, "BTA-$btaVersionId")
        tryDownloadIcon(instance)
    }

    @Throws(IOException::class)
    fun runCatching() {
        removeOldClient()
        val btaVersionId = "bta-${mBtaVersion.versionName}"
        createJson(btaVersionId)
        createProfile(btaVersionId)
    }

    companion object {
        private const val BASE_JSON = "{\"inheritsFrom\":\"b1.7.3\",\"mainClass\":\"net.minecraft.client.Minecraft\",\"libraries\":[{\"name\":\"bta-client:bta-client:%1\$s\",\"downloads\":{\"artifact\":{\"path\":\"bta-client/bta-client-%1\$s.jar\",\"url\":\"%2\$s\"}}}],\"id\":\"%3\$s\"}"
    }
}
